architecture/2024-07-user-defned-types-schema-design.md

```diff
@@ -0,0 +1,305 @@
+  name: 'sample-db'
+  properties: {
+    size: 'L' // Bicep editor can validate this field
+  }
+}
+```
+
+## Design
+
+When user executes a `rad resource-type create` command, the cli should validate the schema object provided in the manifest and report any errors back to the users before making the API call. 
+
+The server side, upon receiving a resource payload for a resource-type resource, must again validate the schema before it saves the resource into database.
+
+We should implement validations on both client and server since rad cli need not be the only client. It is important to validate payloads since these can be any arbitrary data.
+
+### High Level Design
+
+We expect users will provide the structural schema for new resource types and describe it use using [Open API v3](https://github.com/OAI/OpenAPI-Specification/blob/main/versions/3.0.2.md#schema-object). Structural schema ensures explicit type definition for properties and enforces strict validation rules.
+
+### Architecture Diagram
+
+NA 
+
+### Detailed Design
+
+#### Structure
+
+Users provide us with the schema of UDT by defining `properties` in `openAPIV3Schema`. `openAPIV3Schema` is an `object` that holds properties of UDT. There is no support for defining custom fields outside of `properties`.
+
+Each of the property is a user defined property of the UDT. It has a type and optional description and format. The type can be 
+
+* scalar
+
+We support all of the scalars supported by OpenAPI. 
+https://github.com/readmeio/oas-examples/blob/a331c65623f795af68602dd6f02e116f905d9297/3.0/yaml/schema-types.yaml details several examples covering the options available.
+  
+```
+schema: 
+  openAPIV3Schema:
+    type: object
+    properties:
+      size:
+        type: string  
+        description: The size of database to provision
+        enum:
+        - S
+        - M
+        - L
+        - XL
+      host:
+        type: string
+        description: hostname
+        maxLength: 20
+```
+
+* array
+
+`array` must specify a `type` for item
+
+```
+schema:
+  openAPIV3Schema: 
+    type: object
+    properties:
+      ports:
+        type: array
+        description: "ports this resource binds to"
+        item:
+          type: integer
+          format: uint32
+```
+
+* map
+
+We support map through `adddiionalProperties`. This is useful when the resource type allows for dynamic(user defined) keys. We still must specify a type for the value of the property.
+
+```
+schema:
+  openAPIV3Schema:
+    type: object
+    properties:
+      name:
+        type: string
+        description: The name of the resource
+      labels:
+        type: object
+        description: A map of labels for the resource
+        additionalProperties:
+          type: string
+```
+
+
+# Scalars
+
+Schemas will support every construct OpenAPI provides for a scalar field. That includes all of the modifiers like `enum` and `format` and also all of the validation attributes like `minLength`.
+
+# Maps
```

Comment:
An example would be nice. Is it string:string type of map? What does the type of additionalProperties represent? Type of the key or the value of the map?